testutil-server/src/main/java/io/spine/examples/shareaware/testing/server/e2e/AsyncObserver.java

```diff
@@ -0,0 +1,135 @@
+ */
+public class AsyncObserver<S, C> {
+
+    private final Consumer<C> howToCommand;
+
+    private CompletableFuture<S> future = new CompletableFuture<>();
+
+    private S state = null;
+
+    private ObservationState observationState = null;
+
+    /**
+     * Creates the new instance of the {@code AsyncObserver}.
+     *
+     * @param howToObserve
+     *         a callback that defines how to observe the state {@link S}.
+     * @param howToCommand
+     *         a callback that defines how to send a command {@link C}.
+     */
+    public AsyncObserver(
+            Consumer<Consumer<S>> howToObserve,
+            Consumer<C> howToCommand) {
+        this.howToCommand = howToCommand;
+        howToObserve.accept(value -> {
+            state = value;
+            if (future.isDone()) {
+                future = new CompletableFuture<>();
+            }
+            future.complete(value);
+            observationState = ObservationState.OBSERVED;
+        });
+    }
+
+    /**
+     * Posts a command and waits for an update of the entity state that
+     * should occur as a consequence of the posted command.
+     *
+     * <p>If an update of the entity state is not received within 10 seconds,
+     * a {@code TimeoutException} is thrown.
+     */
+    public S onceUpdatedAfter(C command) {
+        howToCommand.accept(command);
+        return waitForUpdate();
+    }
+
+    /**
+     * Returns the current state of the entity if it exists, null otherwise.
+     */
+    public @Nullable S state() {
+        return state;
+    }
+
+    /**
+     * Waits for an update of the entity state to arrive and return this state.
+     *
+     * <p>An update of the entity state should be received within 10 seconds,
+     * otherwise a {@code TimeoutException} will be thrown.
+     */
+    private S waitForUpdate() {
+        if (observationState != null &&
+                observationState != ObservationState.OBSERVED) {
+            future = new CompletableFuture<>();
+        }
+        try {
+            S updatedEntity = future.whenComplete((value, error) -> {
+                                        if (error != null) {
+                                            throw illegalStateWithCauseOf(error);
+                                        }
+                                    })
+                                    .orTimeout(10, SECONDS)
+                                    .get();
+            observationState = ObservationState.UPDATED;
+            return updatedEntity;
+        } catch (InterruptedException | ExecutionException e) {
+            throw illegalStateWithCauseOf(e);
+        }
+    }
+
+    /**
+     * Represents the states of the entity observation.
+     *
+     * <p>The `AsyncObserver` can observe asynchronous changes of the entity's state,
+     * and these states were introduced to restrict the order of the observing operations.
+     */
+    private enum ObservationState {
+        OBSERVED,
```

Comment:
This name is confusing. Can you please document each enum member, so that I could come up with alternative name? 

Please think about this one too.